Use upstream 1.7.1 rest-catalog-open-api.yaml instead.

HonahX opened the pull request:
<!--
    Possible security vulnerabilities: STOP here and contact security@apache.org instead!

    Please update the title of the PR with a meaningful message - do not leave it "empty" or "generated"
    Please update this summary field:

    The summary should cover these topics, if applicable:
    * the motivation for the change
    * a description of the status quo, for example the current behavior
    * the desired behavior
    * etc

    PR checklist:
    - Do a self-review of your code before opening a pull request
    - Make sure that there's good test coverage for the changes included in this PR
    - Run tests locally before pushing a PR (./gradlew check)
    - Code should have comments where applicable. Particularly hard-to-understand
      areas deserve good in-line documentation.
    - Include changes and enhancements to the documentation (in site/content/in-dev/unreleased)
    - For Work In Progress Pull Requests, please use the Draft PR feature.

    Make sure to add the information BELOW this comment.
    Everything in this comment will NOT be added to the PR description.
-->
Depends on #935 

This is a second follow-up of #906. This PR makes the `rest-catalog-open-api.yaml` identical to the upstream released version (1.7.1)

HonahX commented on spec/rest-catalog-open-api.yaml:
```diff
@@ -242,8 +236,6 @@ paths:
         5XX:
           $ref: '#/components/responses/OAuthErrorResponse'
-      security:
-        - BearerAuth: []
```
It seems in Polaris the `/v1/oauth/tokens` has this additional security scheme override. This change is included in the original commit of Polaris.

To ensure this PR does not change anything unexpectedly, I extracted this one to `polaris-catalog-apis/oauth-tokens-api.yaml`. We will likely remove or update this soon anyway according to: https://github.com/apache/polaris/issues/12

dimas-b replied:
Now the `/v1/oauth/tokens` endpoint is defined in twp files, right? What controls which of the definitions is the one used in runtime?

HonahX replied:
Hi @dimas-b, thanks for reviewing. Currently we have `polaris-catalog-service.yaml` as the root yaml for building catalog api service

https://github.com/apache/polaris/blob/84ee2e8f2c728675f81fc6723d93dd2acb5bef6f/spec/polaris-catalog-service.yaml#L76-L79

It uses `$ref` to control the one used in the runtime. This PR updates the `$ref` so that it will point to the `oauth-tokens-api.yaml`

HonahX commented on spec/polaris-catalog-service.yaml:
```diff
@@ -116,6 +118,15 @@ paths:
+  # Not implemented in Polaris
+
+  # /v1/{prefix}/namespaces/{namespace}/tables/{table}/tasks:
+  # Not implemented in Polaris
```
These endpoints are included in 1.7.1 spec but Polaris hasn't supported them yet. I put them here so that we won't lose track.

dimas-b replied:
Should we file GH issues instead, maybe?

jackye1995 replied:
+1 for adding GitHub issues

HonahX replied:
Created one: https://github.com/apache/polaris/issues/966

HonahX commented on spec/rest-catalog-open-api.yaml:
```diff
@@ -221,12 +221,6 @@ paths:
         Clients may also use the token exchange flow to refresh a token that is about to expire by
```
As Jack suggested in https://github.com/apache/polaris/pull/906#discussion_r1936122001, it is better to name this file `iceberg-rest-catalog-open-api.yaml`. I could do this in a follow-up one to make it easier for reviewing

dimas-b replied:
Renames are usually handled fine by GH

HonahX replied:
Cool! Then let me try it and update the PR.

adutra commented on spec/polaris-catalog-apis/oauth-tokens-api.yaml:
```diff
@@ -0,0 +1,99 @@
+#
```
Not a big deal but I'd suggest renaming this file to `polaris-token-service.yaml` to align with other files and also with how we call this endpoint generally in Polaris code. Wdyt?

jackye1995 replied:
I thought the convention is `<some-name>-api.yaml` to refer to APIs in `polaris-catalog-apis` folder, and `polaris-catalog-service.yaml` is outside of that folder is kind of the only exception?

jackye1995 commented on spec/polaris-catalog-apis/oauth-tokens-api.yaml:
```diff
@@ -0,0 +1,99 @@
+              $ref: '../rest-catalog-open-api.yaml#/components/schemas/OAuthTokenRequest'
+      responses:
+        200:
+          $ref: '../rest-catalog-open-api.yaml#/components/responses/OAuthTokenResponse'
```
I think the `OAuthTokenResponse` will also go away once the IRC token endpoint is deprecated, and we should not depend on that

HonahX replied:
Good point! I've extracted all relevant components to `oauth-tokens-api.yaml`

dimas-b commented on spec/polaris-catalog-apis/oauth-tokens-api.yaml:
```diff
@@ -0,0 +1,99 @@
+
+---
+paths:
+  # The /v1/oauth/tokens endpoint is sourced from rest-catalog-open-api.yaml
```
Could we reference elements of `spec/rest-catalog-open-api.yaml` instead and have the `CODE_COPIED_TO_POLARIS` annotation only on `rest-catalog-open-api.yaml` (which is the file actually copied 1:1)?

HonahX replied:
IMHO, the goal here is to extract content outside of IRC to remove the dependency on rest-catalog-open-api.yaml (specifically for v1/oauth/tokens). This gives Polaris the flexibility to update or remove the deprecated version anytime, especially when upgrading IRC to a future release where the token endpoint is removed. I'm +1 on copying the content and adding the `CODE_COPIED_TO_POLARIS` header. WDYT?

dimas-b replied:
SGTM, but we _also_ need `CODE_COPIED_TO_POLARIS` in `rest-catalog-open-api.yaml`, right?

HonahX replied:
Yes, I've added the header to all related files and updated the LICENSE

HonahX commented:
This PR is ready for review since https://github.com/apache/polaris/pull/953 and #935 are in. @jackye1995 @dimas-b @adutra Could you please take another look at this when you have a chance? Thanks!

dimas-b commented on LICENSE:
```diff
@@ -215,7 +215,10 @@ License: https://www.apache.org/licenses/LICENSE-2.0
 
-* spec/rest-catalog-open-api.yaml
+* spec/iceberg-rest-catalog-open-api.yaml
+* spec/polaris-catalog-service.yaml
```
no, this is Polaris' own file, it is NOT from Apache Iceberg, right?

HonahX replied:
Thanks for all the great catch here! It seems I have misunderstood the intended use of the CODE_COPIED_TO_POLARIS flag and the corresponding licensing requirements. My current understanding is that this flag should only be applied to files where all content is copied from another project. Could you please confirm if that’s correct?

Based on this understanding, here’s the current status:
- `iceberg-rest-catalog-open-api.yaml`: identical to upstream released spec. **_Need_** `CODE_COPIED_TO_POLARIS`
- `polaris-catalog-apis/oauth-tokens-api.yaml`: everything is copied from IRC spec's token endpoint definition. **_Need_** `CODE_COPIED_TO_POLARIS`
- `polaris-catalog-service.yaml` - only the server and security definition are copied from IRC spec, while the rest are not. **_Do Not Need_** `CODE_COPIED_TO_POLARIS`
- `bundled-polaris-catalog-service.yaml` - generated yaml that contains both IRC spec and polaris defined ones. _**Do Not Need**_ `CODE_COPIED_TO_POLARIS`

dimas-b replied:
> My current understanding is that this flag should only be applied to files where all content is copied from another project. Could you please confirm if that’s correct?

That's my understanding too :)

dimas-b replied:
> polaris-catalog-service.yaml - only the server and security definition are copied from IRC spec, while the rest are not. Do Not Need CODE_COPIED_TO_POLARIS

I agree. Plus we have a general notice with Iceberg attribution. I think that's fine.

@jbonofre  WDYT?

dimas-b commented on LICENSE:
```diff
@@ -215,7 +215,10 @@ License: https://www.apache.org/licenses/LICENSE-2.0
+* spec/iceberg-rest-catalog-open-api.yaml
+* spec/polaris-catalog-service.yaml
+* spec/polaris-catalog-apis/oauth-tokens-api.yaml
+* spec/generated/bundled-polaris-catalog-service.yaml
```
generated files cannot possibly be copied from Iceberg.

dimas-b commented on spec/generated/bundled-polaris-catalog-service.yaml:
```diff
@@ -14,6 +14,9 @@
 #  specific language governing permissions and limitations
 #  under the License.
+
+# CODE_COPIED_TO_POLARIS
```
This is generate file, it is certainly not copied from other repos.

dimas-b commented on spec/polaris-catalog-service.yaml:
```diff
@@ -17,6 +17,8 @@
 # under the License.
 #
 
+# CODE_COPIED_TO_POLARIS
```
Is this file really copied from Iceberg? It looks like a genuine Polaris source file to me :thinking:

dimas-b commented:
@snazy @RussellSpitzer @jackye1995  @collado-mike : TM, but since it's an API change - please also have a look :)

flyrain commented on spec/polaris-catalog-apis/oauth-tokens-api.yaml:
```diff
@@ -0,0 +1,279 @@
+---
+paths:
+  # The /v1/oauth/tokens endpoint is sourced from iceberg-rest-catalog-open-api.yaml
+  /v1/oauth/tokens:
```
Sorry to miss the context, why do we copy this from iceberg-rest-catalog-open-api.yaml?

HonahX replied:
The original related discussion is here: https://github.com/apache/polaris/pull/906#discussion_r1936132389, https://github.com/apache/polaris/pull/936#discussion_r1941643769.

Since the token endpoint is deprecated and will soon be removed from the upstream IRC spec, we want to extract it from the spec. This allows us to control its removal timeline, update the endpoint as needed, or continue supporting it temporarily for compatibility. [Issue #12](https://github.com/apache/polaris/issues/12).

flyrain commented on spec/polaris-catalog-apis/oauth-tokens-api.yaml:
```diff
@@ -0,0 +1,279 @@
+# under the License.
+#
+
+# CODE_COPIED_TO_POLARIS
```
Should we also give a version like this? Given that future Iceberg rest spec may not even have it.
```
# CODE_COPIED_TO_POLARIS

# Apache Iceberg Version: 1.7.1

```

dimas-b replied:
I agree that in the current form `CODE_COPIED_TO_POLARIS` is too vague.

I'd argue for dependency-specific labels, e.g. `CODE_COPIED_FROM_ICEBERG: 1.7.1`

Certainly not for this PR, but for follow-up change.. if people are in agreement :)

HonahX replied:
Thanks for the suggestions! I've added the comment but I will leave this thread open to continue the discussion on dependency-specific labels : )

flyrain commented on spec/polaris-catalog-apis/oauth-tokens-api.yaml:
```diff
@@ -0,0 +1,279 @@
+
+---
+paths:
+  # The /v1/oauth/tokens endpoint is sourced from iceberg-rest-catalog-open-api.yaml
```
Maybe we could add a comment like this
```suggestion
  # The /v1/oauth/tokens endpoint is copied from iceberg-rest-catalog-open-api.yaml, ensuring that its deprecation and removal schedule is independent of the Iceberg release cycle.
```

HonahX commented:
Thanks everyone for reviewing! I've updated the PR. I saw currently Merging is blocked because 
```
This branch must not contain merge commits
```
but I've cleaned up/rebased commits. Does it mean I have to close and re-open this one or open a new PR?
